distributed/tests/test_stories.py

```diff
@@ -0,0 +1,170 @@
+            (key, "waiting", "processing", {}),
+            (key, "processing", "memory", {}),
+        ],
+    )
+
+    await c.close()
+
+    stories_after_close = s.story(key)
+    assert len(stories_after_close) > len(stories)
+
+    stimulus_ids = {s[-2] for s in stories_after_close}
+    # One more event
+    # - Forget / Release / Free since client closed
+    assert len(stimulus_ids) == 3
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_scheduler_story_stimulus_retry(c, s, a):
+    def task():
+        assert dask.config.get("foo")
+
+    with dask.config.set(foo=False):
+        f = c.submit(task)
+        with pytest.raises(AssertionError):
+            await f
+
+    with dask.config.set(foo=True):
+        await f.retry()
+        await f
+
+    story = s.story(f.key)
+    stimulus_ids = {s[-2] for s in story}
+    # Four events
+    # - Compute
+    # - Erred
+    # - Compute / Retry
+    # - Success
+    assert len(stimulus_ids) == 4
+
+    assert_story(
+        story,
+        [
+            # Erred transitions via released
+            (f.key, "processing", "erred", {}),
+            (f.key, "erred", "released", {}),
+            (f.key, "released", "waiting", {f.key: "processing"}),
+        ],
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_client_story(c, s, a):
+    f = c.submit(inc, 1)
+    assert await f == 2
+    story = await c.story(f.key)
+
+    # Every event should be prefixed with it's origin
+    # This changes the format compared to default scheduler / worker stories
+    prefixes = set()
+    stripped_story = list()
+    for msg in story:
+        prefixes.add(msg[0])
+        stripped_story.append(msg[1:])
+    assert prefixes == {"scheduler", a.address}
+
+    assert_valid_story(stripped_story, ordered_timestamps=False)
+
+    # If it's a well formed story, we can sort by the last element which is a
+    # timestamp and compare the two lists.
+    assert sorted(stripped_story, key=lambda msg: msg[-1]) == sorted(
+        s.story(f.key) + a.story(f.key), key=lambda msg: msg[-1]
+    )
+
+
+class WorkerBrokenStory(Worker):
+    async def get_story(self, *args, **kw):
+        raise CommClosedError
+
+
+@gen_cluster(client=True, Worker=WorkerBrokenStory)
+@pytest.mark.parametrize("on_error", ["ignore", "raise"])
+async def test_client_story_failed_worker(c, s, a, b, on_error):
+    f = c.submit(inc, 1)
+    coro = c.story(f.key, on_error=on_error)
+    await f
+
+    if on_error == "raise":
+        with pytest.raises(CommClosedError):
+            await coro
+    elif on_error == "ignore":
+        story = await coro
+        assert story
+        assert len(story) > 1
+    else:
+        raise ValueError(on_error)
+
+
+@gen_cluster(client=True)
+async def test_worker_story_with_deps(c, s, a, b):
+    """
+    Assert that the structure of the story does not change unintentionally and
+    expected subfields are actually filled
+    """
```

Comment:
This test has been moved from test_worker.py I added it a while ago because the stories where actually broken. Specifically, the recommendations where not there because at some point we're popping off the recommendations dict. In this case, I consider _one_ test that asserts how a worker story looks like specifically to be valuable. Since I introduced it, I don't think we had to touch it often

Reply:
> And do you care about these specific stories? Or are we just copying our current state?

Something in between. I think this test serves a few purposes

1. Make sure that stories of a worker actually work. That's the trivial simple case. Verify it's format, verify how many stimulus IDs are in there (yes, I care about the exact number in this case), etc. I think this test is unique and the only place where we're doing this for the worker.
2. There was a bug where dependencies where not included in the _recommendations_ part of the transition log. To capture this, we could transition to a property based testing approach where I generate a story, pick the 4th element of a message, if it exists, and assert that it's not empty.

The above two points would better be served by property based testing, e.g. what `assert_valid_story` does, with a few additions. There are a few more reasons why this test is more explicit

3. I consider good tests to be part of developer documentation. I consider this test readable and a developer can have a look and see how a transition log actually looks like. What information is in there, what types, what kind of stati do we log, etc. This shows a very simple computation with two keys so every developer should be able to follow this and reason about it to get a better understanding about how a transition log is working.
4. The last and most important reason why I care about the exact story in this example is admittedly a bit of a code smell / anti-pattern. This example shows the exact happy path evolution of a trivial flow through the worker state machine. Strictly speaking, this is a _unit_ test of the worker state machine as well so we're mixing two concerns into one test. This part of the test should probably rather live in `test_worker_state_machine.py` but it is obviously very hard to define a sharp boundary.

> This makes it seem to me like we're testing that stories themselves work

Yes, that's my intention. There are a few use cases where we need stories to test dask functionalities but they are rare, and should be rare, e.g. to protect us from specific race conditions if there is no other feasible way to test it. In these cases, they are required because we don't have a better tool at the moment. As long as these tests that rely on stories to test functionality remain rare, I'm fine with that. For me, this is a trade off between development velocity, technical debt (as in brittle tests) and protection from regressions.

Reply:
Historically when faced with a challenge of how to test very specific behavior I've always tried to find a way around doing this.  Dask has hundreds of tests for very specific situations.  I think that it would be unfortunate if these all used transition logs or something similar.  I think that we would not be able to move forward today if I had made that choice historically.  

Stories are nicer than raw transition logs, I'll give you that.  Maybe they're ok-enough?  I don't know.  I'm concerned because it's so attractive to go down that path, but then developers 1-2 years from now pay the price.

Reply:
> I consider good tests to be part of developer documentation.

I'm fine with that for developers of the story feature.  However I'm not ok with that if this test will be sensitive to changes in other parts of the codebase.

If the answer here is "but we want developers to understand stories" then fine, I can accept that, but we should then have excellent documentation on stories and force all developers to go through it.  I don't view this feature as being at that level of importance.  

Tests as developer documentation can make sense if the tests are isolated.  A developer will only run into a test if they're working on that part of the codebase.  Presenting developers with failing tests randomly selected from the codebase is not an efficient way to train people.

Reply:
So, I'm cool with this test (or any very precise test) if it's unlikely that developers working in other parts of the codebase will trigger this test.  Do we think that that's the case today?

Reply:
I had a look at how often this test (and why) it needed to be changed so far.

We touched it nine times already. With one very small exception, I consider the test valueable for all changes. We might not need it in its current state forever but right now I think it's more useful than a problem

## Functionality around stories themselves
These are changes the test is actually supposed to test
- assert_story tooling c4de83ff35966a1a86f4e23a705fc7dcdbd65683
- write the actual basic test 0a014dac4a1edb090ee17027fabcd41cdd015553
- Refactoring of the test itself 0f2674b61f27a87a779646009b7b1e936d28ef16
- Story format changed 03647c7d946cc47138b2dec223465a7cea77360a

## Major refactorings (impacting stories as well)

- Major refactoring where this test was critical a8d4ffa156cac22b05b9cef3a0599a08e9945146
- Another big refactoring where this was important f6b2e03cb659f19a516eb14c3270f9430a661af3

## Functional test that should be tested elsewhere but is visible in this test
These changes fix major bugs and the fix has an impact in the story. The story should not be used as the primary test source and indeed there are dedicated tests for the regressions themselves.

- Fix a deadlock that can be tested by asserting on unique stim IDs 1670cf85b815a7f547fc040f3f74a24c679dbded
- Fix a major bug that changes the story. There is a proper test for the functionality as well 833c5f6c040feaa4550fa343d8e6e4feef3f84d5 

## Noisy changes due to refactoring / story format changes, etc.

- 5e5b9835a5e6929bb97acf94f0a2eecccc0b6a2f

<details>

<summary> git log -L1736,+42:distributed/tests/test_worker.py -s </summary>

```
commit f6b2e03cb659f19a516eb14c3270f9430a661af3
Author: crusaderky <crusaderky@gmail.com>
Date:   Fri Apr 8 21:00:31 2022 +0100

    Migrate ensure_executing transitions to new WorkerState event mechanism - part 2 (#6062)

commit 0f2674b61f27a87a779646009b7b1e936d28ef16
Author: crusaderky <crusaderky@gmail.com>
Date:   Mon Apr 4 17:54:24 2022 +0100

    Harden test_abort_execution_to_fetch and more (#6026)
commit f6b2e03cb659f19a516eb14c3270f9430a661af3
Author: crusaderky <crusaderky@gmail.com>
Date:   Fri Apr 8 21:00:31 2022 +0100

    Migrate ensure_executing transitions to new WorkerState event mechanism - part 2 (#6062)

commit 0f2674b61f27a87a779646009b7b1e936d28ef16
Author: crusaderky <crusaderky@gmail.com>
Date:   Mon Apr 4 17:54:24 2022 +0100

    Harden test_abort_execution_to_fetch and more (#6026)
commit f6b2e03cb659f19a516eb14c3270f9430a661af3
Author: crusaderky <crusaderky@gmail.com>
Date:   Fri Apr 8 21:00:31 2022 +0100

    Migrate ensure_executing transitions to new WorkerState event mechanism - part 2 (#6062)

commit 0f2674b61f27a87a779646009b7b1e936d28ef16
Author: crusaderky <crusaderky@gmail.com>
Date:   Mon Apr 4 17:54:24 2022 +0100

    Harden test_abort_execution_to_fetch and more (#6026)

commit c4de83ff35966a1a86f4e23a705fc7dcdbd65683
Author: crusaderky <crusaderky@gmail.com>
Date:   Thu Dec 16 17:08:23 2021 +0000

    assert_worker_story (#5598)

commit 03647c7d946cc47138b2dec223465a7cea77360a
Author: Florian Jetter <fjetter@users.noreply.github.com>
Date:   Thu Nov 18 22:45:16 2021 +0100

    Resolve `KeyError`-related deadlock (#5525)

commit 1670cf85b815a7f547fc040f3f74a24c679dbded
Author: Florian Jetter <fjetter@users.noreply.github.com>
Date:   Fri Oct 22 19:44:35 2021 +0200

    Ensure resumed flight tasks are still fetched (#5426)

commit a8d4ffa156cac22b05b9cef3a0599a08e9945146
Author: Florian Jetter <fjetter@users.noreply.github.com>
Date:   Mon Sep 27 14:05:23 2021 +0200

    Worker state machine refactor (#5046)

    Co-authored-by: crusaderky <crusaderky@gmail.com>

commit 5e5b9835a5e6929bb97acf94f0a2eecccc0b6a2f
Author: Florian Jetter <fjetter@users.noreply.github.com>
Date:   Fri Jun 11 19:12:02 2021 +0200

    Forget erred tasks // Fix deadlocks on worker (#4784)

    * Add a test about expected task states in an exception case

    * Minor refactoring about who_has state transitions

    * Remvoe worker kwarg from gather_dep fetch transition

    * Add special case for inflight tasks without dependents

    * Fix compute deadlocks

    * Code review comments

    * Improve request X keys for task debug log

    * Before calling missing-data ensure key is removed from who_has

    * review comments

    * Add test case for same-host missing deps

    * Add test about superfluous data race condition

commit 833c5f6c040feaa4550fa343d8e6e4feef3f84d5
Author: Florian Jetter <fjetter@users.noreply.github.com>
Date:   Tue May 25 12:52:57 2021 +0200

    Ensure busy workloads properly look up who_has (#4793)

commit 0a014dac4a1edb090ee17027fabcd41cdd015553
Author: Florian Jetter <fjetter@users.noreply.github.com>
Date:   Wed May 5 11:15:32 2021 +0200

    Ensure deps are actually logged in worker (#4753)

    * Ensure deps are actually logged in worker

    * Log only keys for gather-dependency log
```

</details>

Reply:
If this is already in main then I don't need to hold this up.  However, I would like for us to pause on adding any more tests like this until we come to some agreement on these style of tests.